### PR TITLE
test: add unit tests for edge:drivers:package command

### PR DESCRIPTION
**Pull request by rossiam**
* refactored some code out of command itself to aid in unit testing
* completed unit tests for the `edge:drivers:package` command and associated code

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] I have read the **[CONTRIBUTING](../CONTRIBUTING.md)** document
- [X] My code follows the code style of this project (`npm run lint` produces no warnings/errors)
- [ ] Any required documentation has been added
- [X] I have added tests to cover my changes

### CODE REVIEW

File: test/unit/lib/commands/drivers/package-utils.test.ts [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+import * as fileUtil from '../../../../../src/lib/file-util'
+import { buildTestFileMatchers, getProjectDirectory, processConfigFile, processFingerprintsFile,
+	processProfiles, processSrcDir } from '../../../../../src/lib/commands/drivers/package-util'
+
+
+jest.mock('fs')
+jest.mock('js-yaml')
+jest.mock('picomatch')
+
+describe('package', () => {
```
Comment (john-u) [inline test/unit/lib/commands/drivers/package-utils.test.ts:16]: ~~package~~  `package-utils`?

File: test/unit/lib/commands/drivers/package-utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+			expect(readdirSyncMock).toHaveBeenCalledTimes(1)
+			expect(readdirSyncMock).toHaveBeenCalledWith('src dir')
+			expect(isDirSpy).toHaveBeenCalledTimes(2)
+			expect(isDirSpy).toHaveBeenCalledWith('src dir/init.lua')
+			expect(isDirSpy).toHaveBeenCalledWith('src dir/test.lua')
+			expect(zipFileMock).toHaveBeenCalledTimes(1) // NOT 2! :-)
+			expect(zipFileMock).toHaveBeenCalledWith('src/init.lua', readStreamMock)
+		})
+
+		it('throws error if nesting is too deep', () => {
```
Comment (john-u) [inline test/unit/lib/commands/drivers/package-utils.test.ts:280]: It would be helpful to have a comment here pointing out that nesting starts at 2 in `processSrcDir`. It was confusing to see limit of 10 but then 2 for loops 1<=8 and 0<=8.

File: src/commands/edge/drivers/package.ts [hunk 3]
```diff
@@ -258,19 +108,21 @@ $ smartthings edge:drivers:package -u driver.zip`]
 			} catch (error) {
 				if ((error as { code?: string }).code === 'ENOENT') {
 					this.log(`No file named "${flags.upload}" found.`)
+				} else {
+					throw error
```
[the remark this answers is not in the document]
Comment (rossiam) [reply src/commands/edge/drivers/package.ts:112]: I noticed it while I was looking at the code so I guess not technically but I think it counts. 😄

File: src/lib/commands/drivers/package-util.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+
+// Utility methods specific to the `edge:drivers:package` command. Split out here to make
+// unit testing easier.
+
+export interface YAMLFileData {
+	[key: string]: string
+}
+
+export const getProjectDirectory = (projectDirectoryFromArgs: string): string => {
```
[the remark this answers is not in the document]
Comment (rossiam) [reply src/lib/commands/drivers/package-util.ts:17]: I actually pondered that name a lot, but not that reason you did. 😄  I don't like the `get` part of it since it's really  determining the name to use and removing a trailing slash if necessary. Perhaps the function is overloaded but it's so small already.

If I renamed it though (which I'm not opposed to, since it is the project directory name we are dealing with after all), I would want to rename the parameter as well to `projectDirNameFromArgs` since that parameter is returned exactly as the output if it's defined and doesn't end in a slash. That is to say, this input and the output _are_ the same with the exception of the input being the optional one from the args and the output is the (potentially) cleaned-up or default version to use.

[the remark this answers is not in the document]
Comment (rossiam) [reply src/lib/commands/drivers/package-util.ts]: I did rename both the parameter and the function as well as the name of the internal variable that is used for the return value to include `name`.

The more I look at it, though, the more I just generally dislike this function. It also asserts the directory exists so it's really doing three different things which is why it's so hard to name.

File: test/unit/commands/edge/drivers/package.test.ts [hunk 5]
```diff
@@ -1,179 +1,284 @@
...
 
 	afterEach(() => {
 		jest.clearAllMocks()
 	})
 
-	const mockExistsAndIsFile = (): void => {
-		existsSyncMock.mockReturnValueOnce(true)
-		const statsIsFileMock = jest.fn().mockReturnValueOnce(true)
-		lstatSyncMock.mockReturnValueOnce({ isFile: statsIsFileMock } as unknown as fs.Stats)
+	const mockProjectDirectoryProcessing = (): void => {
```
[the remark this answers is not in the document]
Comment (rossiam) [reply test/unit/commands/edge/drivers/package.test.ts:71]: This might be a good thing to discuss in the parking lot. I've been kind of torn between doing more of these in the setup and keeping them more explicit in the tests. Putting them up at the top is nice because you can reduce duplication but it makes the tests a little harder to read because you have to check multiple places to see if the mock was happening which has caused me some grief in the past.

File: test/unit/commands/edge/drivers/package.test.ts [hunk 6]
```diff
@@ -1,179 +1,284 @@
...
+		expect(generateNodeStreamMock).toHaveBeenCalledTimes(1)
+		expect(generateNodeStreamMock)
+			.toHaveBeenCalledWith({ type: 'nodebuffer', streamFiles: true, compression: 'DEFLATE' })
+		expect(createWriteStreamMock).toHaveBeenCalledTimes(1)
+		expect(createWriteStreamMock).toHaveBeenCalledWith('driver.zip')
+		expect(pipeMock).toHaveBeenCalledTimes(1)
+		expect(pipeMock).toHaveBeenCalledWith(writeStreamMock)
+		expect(writeStreamOnMock).toHaveBeenCalledTimes(1)
+		expect(writeStreamOnMock).toHaveBeenCalledWith('finish', expect.any(Function))
+		expect(outputItemMock).toHaveBeenCalledTimes(0)
```
[the remark this answers is not in the document]
Comment (rossiam) [reply test/unit/commands/edge/drivers/package.test.ts:121]: Good idea. Fixed.

File: test/unit/lib/commands/drivers/package-utils.test.ts (hunk 1 above)
Comment (rossiam) [reply test/unit/lib/commands/drivers/package-utils.test.ts]: Oops. Fixed.

File: test/unit/lib/commands/drivers/package-utils.test.ts (hunk 2 above)
Comment (rossiam) [reply test/unit/lib/commands/drivers/package-utils.test.ts:280]: Yeah, that is confusing. I added comments to both the code itself and this test.

File: src/lib/commands/drivers/package-util.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rossiam) [reply src/lib/commands/drivers/package-util.ts]: Maybe I should have called it `processProjectDirName`?

[the remark this answers is not in the document]
Comment (rossiam) [reply src/lib/commands/drivers/package-util.ts]: Or print it out and burn it. 😁

[the remark this answers is not in the document]
Comment (john-u) [reply src/lib/commands/drivers/package-util.ts]: After taking a second look, `resolve` may be a good verb to use in place of `get`. But I think it is fine as refactored.

**Review by john-u — APPROVED** [review]

File: src/lib/commands/drivers/package-util.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rossiam) [reply src/lib/commands/drivers/package-util.ts]: I like `resolve` much better than `get`. I've used this and fixed the couple places where the old name was part of another variable name (in tests) that I forgot to fix with my previous rename.
